docs/dsl-language.md

```diff
@@ -0,0 +1,133 @@
+- Local variables use bare identifiers (e.g., `let n = 0;`). Shadowing state field names is not allowed.
+- Initialize state explicitly in `on start { ... }` or during `migrate` steps; avoid relying on implicit defaults.
+
+### Constructor parameters
+- Fiber constructor parameters (identity, handles like queues, config) are read-only fields of the fiber instance.
+- Access them as `self.<param>` inside the fiber (e.g., `self.name`, `self.inbox_queue`).
+- Parameters cannot be reassigned; locals remain bare identifiers. Shadowing parameter names is not allowed.
+- In `select`, the sugar `self.queue.await` is valid and desugars to `await recv(self.queue)`.
+
+### State migrations
+- Syntax: `migrate vN -> vN+1 { /* transforms */ }` placed inside the `fiber` block before the next `state vN+1`.
+- Scope:
+  - `from.<field>`: read-only view of the previous state snapshot (version N).
+  - `self.<field>`: the new state (version N+1) you must initialize.
+- Rules:
+  - Pure only: no `await`, `select`, `send`, or `external` calls inside `migrate`.
+    - that's questionable. Of course pure migrations are easier, but I can easily imagine migration where I need to make an http call for some data
+    - maybe in the first version if external data is needed, use a two-phase pattern: add new fields, deploy code to backfill via normal runtime flows, then finalize with a follow-up migration
+  - Explicit init: every newly introduced or type-changed field in `state vN+1` must be assigned; unchanged fields carry forward implicitly.
+  - Determinism: transformations must be deterministic and terminate.
+  - Type changes: allowed if you provide an explicit transform; otherwise keep the same type.
+  - Collections: when changing key types in `Map`/`Set`, ensure canonical encoding order is preserved by re-encoding keys.
+- Example (direct rename + add default):
+  ```dsl
+  migrate v1 -> v2 {
+    self.count = from.seen;
+    self.last_input = None;
+  }
+
+  state v2 { count: U64, last_input: Option<String> }
+  ```
+
+## Transactions and IDs
+- Each transaction has a unique `UniqueU64BlobId` (from the gateway’s key range).
+- Idempotency by design: re-applying a transaction yields the same result.
+
+## Resource Limits
+- We track CPU/memory/I/O “cost”.
+- Per-transaction and per-fiber limits apply; hit a limit -> get a backpressure.
+
+## Build Pipeline
+- DSL -> typed AST -> Maroon IR -> generated code that the runtime executes.
+- You’ll see types like `Value`, `CreatePrimitiveValue`, `SetPrimitiveValue`, `SelectArm`, `State` in the generated layer.
+- Runtime input/output: `Input = (LogicalTimeAbsoluteMs, Vec<TaskBlueprint>)`, results are `(UniqueU64BlobId, Value)`.
+
+## Static Checks (examples)
+- `pure` functions can’t call effects or timers.
+- `Map`/`Set` keys must be orderable (canonical encoding available).
+- `select` cases must be cancellable or time-bounded.
```

Comment:
Yes, unless they must be single-step by definition, for what can not be cancelled if it's already in flight!